Adding support for project.json files in pack.

toddm opened the pull request:
@emgarten @spadapet

toddm committed:
Adding support for project.json files in pack.

toddm committed:
Adding project.json tests and fixing bugs found by the tests.

toddm committed:
Fixing a unit test.

dnfclas commented:
Hi **@toddm**, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:
```diff
@@ -577,7 +582,7 @@ private void RecursivelyApply(Action<ProjectFactory> action, dynamic alreadyAppl
                 string fullPath = item.GetMetadataValue("FullPath");
                 if (!string.IsNullOrEmpty(fullPath) &&
-                    !NuspecFileExists(fullPath) &&
+                    !NuspecFileExists(fullPath) && !File.Exists(Path.Combine(Path.GetDirectoryName(fullPath), "project.json")) &&
```
nit: you could use the constant for project.json from: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/ProjectJsonPathUtilities.cs#L11

emgarten replied:
This should support projectName.project.json also, take a look at https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/ProjectJsonPathUtilities.cs#L28

toddm replied:
Added support for name.project.json.

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -28,7 +31,8 @@ public class PackCommandRunner
             ".nproj",
             ".btproj",
-            ".dxjsproj"
+            ".dxjsproj",
```
should .xproj be added?

toddm replied:
Done

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -70,7 +74,11 @@ private PackageArchiveReader BuildPackage(string path)
             string extension = Path.GetExtension(path);
 
-            if (extension.Equals(NuGetConstants.ManifestExtension, StringComparison.OrdinalIgnoreCase))
+            if (Path.GetFileName(path).Equals("project.json"))
```
add projectName.project.json support

emgarten replied:
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/ProjectJsonPathUtilities.cs#L127

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -80,6 +88,162 @@ private PackageArchiveReader BuildPackage(string path)
+
+        private static void LoadProjectJsonFile(PackageBuilder builder, string path, string basePath, string id, Stream stream, Func<string, string> propertyProvider)
+        {
+            string content = Preprocessor.Process(stream, propName => propertyProvider(propName));
```
This would be great, but it should be removed until we can get agreement on this.

toddm replied:
Done

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -80,6 +88,162 @@ private PackageArchiveReader BuildPackage(string path)
+            builder.Copyright = spec.Copyright;
+            if (spec.Authors.Any())
+            {
+                builder.Authors.Add(spec.Authors.First());
```
Why only the first author?

toddm replied:
Switched to using all.

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -80,6 +88,162 @@ private PackageArchiveReader BuildPackage(string path)
+                    {
+                        dependencies.Add(new PackageDependency(dependency.Name, dependency.LibraryRange.VersionRange));
+                    }
+                    PackageDependencyGroup group = new PackageDependencyGroup(framework.FrameworkName, dependencies);
```
This should support include/exclude also

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -80,6 +88,162 @@ private PackageArchiveReader BuildPackage(string path)
+                {
+                    var flags = dependency.IncludeType.ToString();
+                    
+                    dependencies.Add(new PackageDependency(dependency.Name, dependency.LibraryRange.VersionRange, flags.Split(new char[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries), new List<string>()));
```
Does this always add include flags? If these match the default the include attribute should not be written to the nuspec.

emgarten replied:
This should also check suppressParent. The suppressParent flags are applied for the consuming project, which in this case is the project that installs the nupkg, so basically the include flags for the nupkg dependency within a nupkg is always: include - exclude - suppressParent.

https://github.com/NuGet/Home/wiki/%5BSpec%5D-Managing-dependency-package-assets

If the include flags are empty (or suppressParent is all) then the dependency should not be added to the nupkg at all. These are called type: build dependencies and they are only used by the project itself, not by any consumers.

toddm replied:
Exclude and suppressParent are both handled in JsonPackageSpecReader so I don't need to worry about them here.  I'm adding logic to skip the dependency if flags == None and to not pass in the flags if they are == All.

---

In reply to: [57965099](https://github.com/NuGet/NuGet.Client/pull/432#discussion_r57965099) [](ancestors = 57965099)

toddm replied:
The project.json parser handles excludes and suppressParent flags, and gives me just the includes left after that.  I added logic to not include flags if they are set to All, and to skip the dependency if set to None.

emgarten commented on src/NuGet.Core/NuGet.LibraryModel/LibraryIncludeFlag.cs:
```diff
@@ -13,6 +13,6 @@ public enum LibraryIncludeFlags : ushort
         ContentFiles = 1 << 4,
         Analyzers = 1 << 5,
-        All = 0xFFFF
+        All = Analyzers | Build | Compile | ContentFiles | Native | Runtime
```
:+1:

emgarten commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs:
```diff
@@ -1513,6 +2000,95 @@ public class Class1
+    ""json"": { }
+  }
+}");
+                var msbuild = Path.Combine(
```
Could this test be optional when msbuild 4.0 doesn't exist (and the others)?

https://github.com/NuGet/Home/issues/2397

toddm replied:
Done.

emgarten commented:
It might be better to set exclude on dependencies instead of include. That way unknown future things can still flow through. Otherwise you risk a situation such as A -> B -> C  where A -> B excludes build items, then years later C is updated to a higher version that contains new assets, but those assets are blocked by the A -> B edge.

emgarten commented:
:watch:

toddm committed:
Fixing up from CR comments

toddm commented:
Switching to excludes instead of includes would require changing the JsonPackageSpecReader since it handles all of that logic.  Do you think it is worth doing that?

I've addressed all of your comments except this one about excludes vs includes and pushed it.

emgarten commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs:
```diff
@@ -1513,6 +1942,100 @@ public class Class1
+                    @"{
+  ""version"": ""1.0.0"",
+  ""authors"": [ ""feiling"" ],
+  ""owners"": [ ""feiling"" ],
```
You can take Fei's name out of these, just use 'author1' or something :smile:

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:
```diff
@@ -82,15 +83,19 @@ public static PackageSpec GetPackageSpec(Stream stream, string name, string pack
             packageSpec.Description = rawPackageSpec.GetValue<string>("description");
             packageSpec.Authors = authors == null ? new string[] { } : authors.ValueAsArray<string>();
             packageSpec.Owners = owners == null ? new string[] { } : owners.ValueAsArray<string>();
+            packageSpec.ContentFiles = contentFiles == null ? new string[] { } : contentFiles.ValueAsArray<string>();
```
Does this work for content files? I would expect it to need more information to get it into the right tfm/langugage folder.

emgarten commented:
@toddm the fixes looks good. 

I don't think the model needs to change, this should be specific to pack since it is based against the nupkg -> nupkg default of LibraryIncludeFlags.NoContent 

We can go through the scenarios offline

spadapet commented on src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:
```diff
@@ -716,7 +727,11 @@ private PackageDependency CreateDependencyFromProject(dynamic project)
+                if (!projectFactory.ProcessJsonFile(builder, project.DirectoryPath, null))
+                {
+                    projectFactory.ProcessNuspec(builder, null);
+                }
```
Add line break after close curly

spadapet commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs:
```diff
@@ -1513,6 +1942,100 @@ public class Class1
+                // Assert
+                Assert.Contains("Package dependencies must specify a version range.", r.Item3);
+            }
+        }
```
Maybe there's a reasonable way to share most of the code between the project.json and the .csproj version of these tests.

toddm replied:
I gave it a try on the first test, but there were a number of small changes between the two that made it pretty ugly.  I think it's fine this way even though there is some duplication.

spadapet replied:
ok

spadapet commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -419,7 +613,7 @@ internal static string GetInputFile(PackArgs packArgs, IEnumerable<string> files
                     goto default;
                 default:
-                    throw new ArgumentException("Please specify a nuspec or project file to use");
+                    throw new ArgumentException("Please specify a nuspec, project.json, project file to use");
```
Doesn't this need to be localizable?

toddm replied:
Good catch - changing it.

---

In reply to: [58121029](https://github.com/NuGet/NuGet.Client/pull/432#discussion_r58121029) [](ancestors = 58121029)

toddm committed:
Another round of updates from CR comments.

toddm commented:
I pushed a new commit to address comments.

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -215,58 +220,44 @@ private static void LoadProjectJsonFile(PackageBuilder builder, string path, str
 
-                    builder.DependencyGroups.Add(group);
+                var flags = dependency.IncludeType.ToString();
+                IReadOnlyList<string> includes = flags.Split(new char[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries);
```
This should remove suppressParent flags also

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -215,58 +220,44 @@ private static void LoadProjectJsonFile(PackageBuilder builder, string path, str
-                    builder.DependencyGroups.Add(group);
+                var flags = dependency.IncludeType.ToString();
+                IReadOnlyList<string> includes = flags.Split(new char[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries);
+                if (dependency.IncludeType == LibraryIncludeFlags.All || includes.SequenceEqual(allIncludeFlags))
```
Matching NoContent would mean there are no flags since that is the default
If it matches all it needs include="all"

This is very optimized for the defaults I guess so the extra work all ends up on pack

emgarten commented on src/NuGet.Core/NuGet.Commands/PackCommandRunner.cs:
```diff
@@ -215,58 +220,44 @@ private static void LoadProjectJsonFile(PackageBuilder builder, string path, str
+                }
+                else
+                {
+                    packageDependencies.Add(new PackageDependency(dependency.Name, dependency.LibraryRange.VersionRange, flags.Split(new char[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries), new List<string>()));
```
This should be inverted and only the excluded flags added to exclude if possible

spadapet commented:
:shipit:

toddm committed:
Changing the dependency include/exclude/suppressParent logic.

toddm commented:
Updated the dependency include/exclude/suppressParent logic again.  Justin, can you see if this lines up with the expected logic?

toddm commented:
Justin signed off offline.  Merging to Dev.
